[3기-C 권성준] JPA 게시판 과제 제출합니다.

Seongjun-Kwon opened the pull request:
## 📌 과제 설명
안녕하세요 마르코님, 명환님! 코드 리뷰 잘 부탁드립니다.
JPA를 이용해 게시판 API 를 구현하고, REST DOCS를 통해 문서화하였습니다.

## 👩‍💻 요구 사항과 구현 내용
- [x]  **SpringDataJPA 를 설정한다.**
    - [x]  datasource : h2 or mysql
- [x]  **엔티티를 구성한다**
    - [x]  회원(User)
        - id (PK) (auto increment)
        - name
        - age
        - hobby
        - **created_at**
        - **created_by** -> **last_modified_by** 로 수정
    - [x]  게시글(Post)
        - id (PK) (auto increment)
        - title
        - content
        - **created_at**
        - **created_by** -> **last_modified_by** 로 수정
    - [x]  회원과 게시글에 대한 연관관계를 설정한다.
        - 회원과 게시글은 1:N 관계이다.
    - [x]  게시글 Repository를 구현한다. (PostRepository)
- [x]  **API를 구현한다.**
    - [x]  게시글 조회
        - 페이징 조회 (GET "/posts")
        - 단건 조회 (GET "/posts/{id}")
    - [x]  게시글 작성 (POST "/posts")
    - [x]  게시글 수정 (PATCH "/posts/{id}")
- [x]  **REST-DOCS를 이용해서 문서화한다.**

## ✅ 프로그램 실행 방법
intellij 에서 아래처럼 DB 환경변수를 등록해주세요.
DB_URL: jdbc:h2:tcp://localhost/~/test
DB_USER: sa
DB_PASSWORD:

## ✅ PR 포인트 & 궁금한 점
1. PostResponseDto 가 User 에 관한 정보를 얼만큼 가지고 있어야하는지에 대한 고민이 있었습니다. 일단 현재는 최소한의 필요한 정보만 주고자 하여 User의 id 값만 가지도록 하였습니다만, 필요에 따라서 id 외의 정보도 필요한 일이 있을 것 같습니다. 그럴 때마다 새로운 Dto 를 생성해주는게 옳은 방식일까요?

2.  테스트 코드를 많이 작성해보았습니다. 잘못되었거나 아쉬운 부분이 있으면 말씀해주시면 감사하겠습니다!

3. Repository 테스트가 필요한가라는 생각을 했습니다. 제가 작성한 코드가 아니고 JPA 를 사용하는 것인데도 테스트가 필요할까요?

Seongjun-Kwon committed:
chore: 프로젝트 생성

Seongjun-Kwon committed:
chore: 프로젝트 초기 설정

Seongjun-Kwon committed:
feat: User, Post 엔티티 구현

Seongjun-Kwon committed:
feat: User 와 Post 1:N 연관관계 설정

Seongjun-Kwon committed:
feat: BaseEntity 구현

- 생성시간, 생성한 사람 필드로 가짐.

Seongjun-Kwon committed:
feat: UserRepository, PostRepository 구현

Seongjun-Kwon committed:
chore: JPA 설정 변경

- ddl-auto: create
- show_sql: true
- format_sql: true
- log 레벨 info 로 변경

Seongjun-Kwon committed:
refactor: Post 엔티티 수정

- 검증 어노테이션 수정
- 생성자 추가 및 수정

Seongjun-Kwon committed:
refactor: User 엔티티 수정

- 검증 어노테이션 수정
- 매핑되는 테이블명 수정

Seongjun-Kwon committed:
test: 엔티티 생성 검증, 연관관계 설정 테스트

Seongjun-Kwon committed:
test: Repository CRUD 테스트

Seongjun-Kwon committed:
feat: Post 관련 DTO 구현

Seongjun-Kwon committed:
feat: PostConverter 구현

- DTO 들과 Entity 사이의 변환 책임을 가짐.

Seongjun-Kwon committed:
feat: Post, User 조회 안될 시 발생하는 Exception 구현

Seongjun-Kwon committed:
feat: PostService 구현

Seongjun-Kwon committed:
refactor: BaseEntity 수정

- createdBy 필드 삭제 후 lastModifiedAt 필드 추가
- Post 는 user 를 가지고, User 는 createdBy 가 필요 없기 때문.

Seongjun-Kwon committed:
fix: db datasource 설정에 hikari 명시해주지 않아서 충돌나던 문제 해결

Seongjun-Kwon committed:
refactor: Post 요청 DTO 에 생성자 추가

Seongjun-Kwon committed:
feat: PostApiController 구현

Seongjun-Kwon committed:
feat: 예외 발생 시 응답 객체 구현

Seongjun-Kwon committed:
feat: ControllerAdvice 구현

Seongjun-Kwon committed:
chore: PostConverter 에 필요 없는 @Component 어노테이션 제거

Seongjun-Kwon committed:
refactor: 패키지 의존성 관리를 위해 DTO 위치를 domain 패키지로 옮김.

- 기존처럼 web 패키지에 DTO 가 위치하면 service 가 web 패키지에 의존하게 되고, 이는 controller -> service -> repository 의존 흐름에 맞지 않음.

Seongjun-Kwon committed:
refactor: Post 엔티티 수정

- Equals(), HashCode() 추가
- 생성자 추가

Seongjun-Kwon committed:
refactor: User 엔티티 수정

- Equals(), HashCode() 추가
- 생성자 추가
- posts 필드에 cascade 옵션 All 로 변경. User 제거 시에 함께 제거되도록 함.

Seongjun-Kwon committed:
refactor: Bean validation 위한 @Validated 어노테이션 추가

Seongjun-Kwon committed:
test: PostConverter 테스트 추가

Seongjun-Kwon committed:
test: PostService 테스트 추가

Seongjun-Kwon committed:
test: PostApiController 테스트 추가

Seongjun-Kwon committed:
docs: REST DOCS 를 통한 문서화 관련 파일 추가

Seongjun-Kwon committed:
docs: index.adoc 내의 snippets 경로 수정

Seongjun-Kwon committed:
feat: User 관련 DTO 구현

Seongjun-Kwon committed:
feat: UserConverter 구현

- DTO 들과 Entity 사이의 변환 책임을 가짐.

Seongjun-Kwon committed:
feat: UserService 구현

Seongjun-Kwon committed:
feat: UserApiController 구현

Seongjun-Kwon committed:
refactor: PostNotFoundException, UserNotFoundException 이 EntityNotFoundException 상속하도록 변경.

Seongjun-Kwon committed:
refactor: Post 관련 DTO 에 검증 실패 시 메세지 추가

Seongjun-Kwon committed:
chore: controller 에 매핑된 url 없을 시 NoHandlerFoundException 예외 발생하도록 설정 변경

Seongjun-Kwon committed:
refactor: @Validated 대신 @Valid 사용

- MethodArgumentNotValidException 사용하여 BindingResult 가져오기 쉽게 하기 위해서.

Seongjun-Kwon committed:
refactor: 예외 처리 수정

- CommonControllerAdvice 이름 GlobalControllerAdvice 로 수정
- PostControllerAdvice 가 담당하던 PostNotFoundException 을 EntityNotFoundException 로 추상화하여서 한번에 처리

Seongjun-Kwon committed:
chore: 패키지 구조 변경

gw0215 commented on src/docs/asciidoc/index.adoc:
```diff
@@ -0,0 +1,41 @@
+==== GET /api/v1/posts/{id}
+
+.Request
+include::{snippets}/post-get/http-request.adoc[]
```
parameter의 내용은 docs에 없네요?

Seongjun-Kwon replied:
인지하지 못했네요. 수정했습니다~

gw0215 commented on src/main/java/com/programmers/jpaboard/domain/BaseEntity.java:
```diff
@@ -0,0 +1,27 @@
+@MappedSuperclass
+@EntityListeners(AuditingEntityListener.class)
+@Getter
+public class BaseEntity {
```
이거 따로 만드는 경우가 있나요?

Seongjun-Kwon replied:
따로 만드는 경우가 없기에 추상 클래스로 변경하여 생성을 제한하였습니다.

gw0215 commented on src/main/java/com/programmers/jpaboard/web/post/PostApiController.java:
```diff
@@ -0,0 +1,56 @@
+	@PostMapping
+	ResponseEntity<PostResponseDto> createPost(@RequestBody @Valid PostCreateRequestDto postCreateRequestDto) {
+		PostResponseDto createdPostDto = postService.createPost(postCreateRequestDto);
+		return ResponseEntity.ok(createdPostDto);
```
200도 좋지만 201(created)도 어떤가요?

Seongjun-Kwon replied:
더 명시적인 것 같습니다. 변경하였습니다.

gw0215 commented on src/test/java/com/programmers/jpaboard/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,115 @@
+		assertThat(savedPostDto.getContent()).isEqualTo(post.getContent());
+		assertThat(savedPostDto.getUserId()).isEqualTo(post.getUser().getId());
+		assertThat(savedPostDto.getCreatedAt()).isEqualTo(post.getCreatedAt());
+		assertThat(savedPostDto.getLastModifiedAt()).isEqualTo(post.getLastModifiedAt());
```
hasFieldOrPropertyWithValue() 사용해 보세요 각각의 장단점은 있을겁니다.

Seongjun-Kwon replied:
변경해서 사용해봤는데 가독성이 훨씬 좋아진 것 같아요!

park-se-jun replied:
저두 배워가요 건우님 감사합니다

gw0215 commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,137 @@
+		mockMvc.perform(post("/api/v1/posts")
+				.contentType(MediaType.APPLICATION_JSON)
+				.content(objectMapper.writeValueAsString(postCreateRequestDto)))
+			.andExpect(status().isOk())
```
상태코드만 검증하고 있는데 body의 내용도 검증해보는건 어떤가요?

Seongjun-Kwon replied:
수정하였습니다.

gw0215 commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,137 @@
+		mockMvc.perform(get("/api/v1/posts")
+				.param("page", String.valueOf(0))
+				.param("size", String.valueOf(2))
+				.contentType(MediaType.APPLICATION_JSON))
```
없어도 테스트가 수행되지 않나요?
왜 없어도 될까요?

Seongjun-Kwon replied:
.perform 메서드에서 지정해준 내용들은 테스트 결과에는 영향을 미치지 않기 때문입니다.
추가적으로 현재 이 메서드에서는 .contentType(MediaType.APPLICATION_JSON)) 이 아니기 때문에 제거하였습니다.

gw0215 commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,137 @@
+		PostResponseDto createdPost = postService.createPost(postCreateRequestDto);
+
+		mockMvc.perform(get("/api/v1/posts/{postId}", createdPost.getId())
+				.contentType(MediaType.APPLICATION_JSON))
```
acceptType을 지정해보면 어떨까요?

gw0215 commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,137 @@
+			.andExpect(status().isOk())
+			.andDo(print())
+			.andDo(document("post-get",
+				responseFields(
```
parameter 내용이 안보입니다!

Seongjun-Kwon replied:
수정하였습니다!

gw0215 reviewed (commented):
성준님 고생하셨습니다.
성준님의 코드를 보고 성장하신게 보입니다. 
간단한 리뷰남깁니다.

Seongjun-Kwon committed:
refactor: BaseEntity 추상클래스로 만들어서 생성을 제한함.

Seongjun-Kwon committed:
refactor: Controller 에서 리소스 생성 성공 시 201 로 응답하게 함.

Seongjun-Kwon committed:
test: 객체 필드 값 테스트 할 때, hasFieldOrPropertyWithValue() 사용하여 가독성 높임.

Seongjun-Kwon committed:
test: PostApiControllerTest 수정.

- 통합 테스트에서 slice 테스트로 변경.
- Response Body 의 Json 필드 값 테스트 추가.
- 불필요한 contentType 헤더 지정 제거.
- accept 헤더 명시해줌.

Seongjun-Kwon committed:
docs: REST DOCS 문서에 request parameter 값 추가.

Seongjun-Kwon commented:
>

park-se-jun commented on src/main/java/com/programmers/jpaboard/common/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import javax.persistence.EntityNotFoundException;
+
+public class UserNotFoundException extends EntityNotFoundException {
```
EntityNotFoundException을 이용해서 커스텀 에러를 만드는거 배워갑니다

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/dto/PostCreateRequestDto.java:
```diff
@@ -0,0 +1,27 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
noArgsConstructor가 필요한지 궁금합니다

Seongjun-Kwon replied:
RequestBody 어노테이션을 통해 역직렬화 시 기본생성자와 getter or setter or public 필드를 통해 property 명을 파악하고 reflect 를 이용하여서 주입해주는 것으로 알고 있어, 기본생성자를 만들어주었습니다.
기본생성자를 통한 외부에서의 무분별한 생성을 막기 위해 protected 접근제한자를 추가하면 좋을 것 같아 추가하였습니다. 
또, response dto 들에는 기본생성자가 필요하지 않기에 제거하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/dto/PostResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
noArgsConstructor가 필요한지 궁금합니다222

Seongjun-Kwon replied:
response dto 들에는 기본생성자가 필요하지 않기에 제거하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/dto/PostUpdateRequestDto.java:
```diff
@@ -0,0 +1,22 @@
+import lombok.NoArgsConstructor;
+
+@Getter
+@NoArgsConstructor
```
noArgsConstructor가 필요한지 궁금합니다3333

Seongjun-Kwon replied:
RequestBody 어노테이션을 통해 역직렬화 시 기본생성자와 getter or setter or public 필드를 통해 property 명을 파악하고 reflect 를 이용하여서 주입해주는 것으로 알고 있어, 기본생성자를 만들어주었습니다.
기본생성자를 통한 외부에서의 무분별한 생성을 막기 위해 protected 접근제한자를 추가하면 좋을 것 같아 추가하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/entity/Post.java:
```diff
@@ -0,0 +1,76 @@
+		setUser(user);
+	}
+
+	public Post(Long id, String title, String content) {
```
user 없는 post 가 있을 수 있나요?

Seongjun-Kwon replied:
setUser 메서드를 통해서 연관관계를 연결해주려고 했는데, 현재 로직에서는 user 없는 post 가 없으니 생성자 파라미터에 user 를 추가하여 생성자 안에서 setUser를 해주면 좋을 것 같습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/service/PostService.java:
```diff
@@ -0,0 +1,65 @@
+	}
+
+	@Transactional(readOnly = true)
+	public Page<PostResponseDto> getPosts(Pageable pageable) {
```
Page를 전부 내리면 api 문서화 시키기에도 힘들고 프론트 엔드 개발자도 너무 많은 필드를 알아야해서 힘들것 같아요😭 중복되는것도 막 있고....

Seongjun-Kwon replied:
List<PostResponseDto> 를 필드로 갖는 Dto 를 만들어서 반환하도록 수정했습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/util/PostConverter.java:
```diff
@@ -0,0 +1,24 @@
+import com.programmers.jpaboard.domain.post.entity.Post;
+import com.programmers.jpaboard.domain.user.entity.User;
+
+public class PostConverter {
```
static class인데 인스턴스가 생성될 수 있을 것 같아요

Seongjun-Kwon replied:
현재는 static 메서드만 있기에 인스턴스 생성을 막아주는 것이 좋은 것 같습니다.
수정하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/post/dto/PostResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+@Getter
+@NoArgsConstructor
+@AllArgsConstructor
+public class PostResponseDto {
```
requestDto는 validation을 했는데 responseDto는 validation을 하지 않은 이유가 있나요??

Seongjun-Kwon replied:
사용자의 입력 값을 담는 requestDto 에서 검증을 하고, DB 와 매핑되는 entity 생성 시에 검증을 해주었습니다.
responseDto 는 entity 를 변환하여 생성하기에 검증이 필요하지 않다고 생각하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/user/dto/UserResponseDto.java:
```diff
@@ -0,0 +1,20 @@
+	private String email;
+	private int age;
+	private String hobby;
+	private List<Long> postIds;
```
postIds가 필요할까요? 요구사항에 따라 다르겠지만, user를 get하는 api를  프론트엔드에서 콜을 할 때 보통 user 관련 정보를 원하는 것이니
1. User 관련 정보만 보여주기
2. User 관련 정보 + user 가 작성한 post들의 정보를 리스트로 보여주기

이렇게 2가지 정도의 데이터를 기대할 것 같은데, postIds 만 보내는건 프론트 엔드에서 원하는 뷰를 만들기에 부족한 정보인것 같아요
차라리 User 관련 정보만 내려서, 프론트엔드에서 get posts/{id} 콜을 한번 더 하게 하거나 List< PostResponseDto > 를 내려주는게 좋지 않을까요??

사실 잘 모르겠는데 성준형님의 생각은 어떤가요?

Seongjun-Kwon replied:
저도 세준님의 의견이 좋은 것 같습니다.
1번, 2번 둘 다 dto 를 따로 만들면 좋을 것 같습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/user/entity/User.java:
```diff
@@ -0,0 +1,82 @@
+	@Size(max = 20)
+	private String hobby;
+
+	@OneToMany(mappedBy = "user", cascade = CascadeType.ALL)
```
cascadeType all로 지정한 이유가 뭔가요??( 내가 cascadeType에 대해 잘 몰라서 물어본거임)

Seongjun-Kwon replied:
편의를 위해 사용자를 영속화했을 때, 관련된 게시글도 함께 영속화하기 위해서입니다.
remove, detach 와 같이 제거하는 것은 사용자가 사라져도 게시글은 남아있을 수 있기에 해주지 않는 것이 좋다고 생각합니다. 
persist 만 해주는 것으로 수정했습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/user/service/UserService.java:
```diff
@@ -0,0 +1,25 @@
+import lombok.RequiredArgsConstructor;
+
+@Service
+@RequiredArgsConstructor
```
이거는 final 이 붙거나 @NotNull이 붙은 필드의 생성자를 자동으로 만들어주는거군요! 좋은거 알아갑니다 😁

park-se-jun commented on src/main/java/com/programmers/jpaboard/domain/user/util/UserConverter.java:
```diff
@@ -0,0 +1,31 @@
+import com.programmers.jpaboard.domain.user.dto.UserResponseDto;
+import com.programmers.jpaboard.domain.user.entity.User;
+
+public class UserConverter {
```
이것도 정적 클래스인데 인스턴스가 생길 수 있을 것 같아요

Seongjun-Kwon replied:
수정하였습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/web/common/GlobalControllerAdvice.java:
```diff
@@ -0,0 +1,43 @@
+import org.springframework.web.bind.annotation.RestControllerAdvice;
+import org.springframework.web.servlet.NoHandlerFoundException;
+
+@RestControllerAdvice
```
에러핸들링용 컨트롤러가 너무 멋져요!! 본받아야 겠습니다 ㅎㅎ 
근데 postMan으로 잘못된 요청을 날려보니 에러발생 메시지가 너무 적나라하게 나오는데 wrapping 하거나 하는건 어떻게 생각하세요?

Seongjun-Kwon replied:
예외를 최대한 커버해서 메시지를 만들어줬다고 생각했는데, NoHandlerFoundException 예외의 경우를 처리해주지 못했었네요 😂
수정했습니다.

park-se-jun commented on src/main/java/com/programmers/jpaboard/web/post/PostApiController.java:
```diff
@@ -0,0 +1,57 @@
+	private final PostService postService;
+
+	@PostMapping
+	ResponseEntity<PostResponseDto> createPost(@RequestBody @Valid PostCreateRequestDto postCreateRequestDto) {
```
제가 항상 고민하던 부분중 하나인데요. create 나 update의 response로 뭐가 가장 좋을까요?? get에서도 사용하는 post의 정보가 담긴 dto를 내려주기로 선정한 특별한 이유가 있을까요?? 프론트엔드에서 그걸로 뭘 할 수 있을까요?? 성공 실패 여부만 알면 되지 않을까요? response body가 있긴 해야할까요?? 아니면 리다이렉트 될 주소(해당하는 post의 url)을 내리는게 좋을까요? postId만 내리는건 어떨까요?? 전 잘 모르겠는데 성준형님의 의견이 궁금합니다.

Seongjun-Kwon replied:
찾아봤는데 cretaed 상태코드로 넘겨준다면 생성된 자원의 주소도 함께 넘겨주는 것이 규칙이 된 것 같습니다. 이에 맞춰서 수정했습니다.

만약 프론트쪽에서 생성된 자원의 주소를 지정해주는 특수한 상황이라면 created 를 사용하지 않고 ok 로 성공했다는 응답과 id 정도만 보내줘도 될 것 같습니다.

https://developer.mozilla.org/en-US/docs/Web/HTTP/Status/201
https://www.tabnine.com/code/java/methods/org.springframework.http.ResponseEntity/created

park-se-jun commented on src/main/java/com/programmers/jpaboard/web/post/PostApiController.java:
```diff
@@ -0,0 +1,57 @@
+	}
+
+	@GetMapping
+	ResponseEntity<Page<PostResponseDto>> getPosts(Pageable pageable) {
```
다른데서도 적었는데 Page를 전부 내리는것 보다는 필요한 것만 내리는것에 대해 어떻게 생각하세요?

Seongjun-Kwon replied:
수정했습니다.

park-se-jun commented on src/test/java/com/programmers/jpaboard/domain/post/entity/PostTest.java:
```diff
@@ -0,0 +1,80 @@
+
+import com.programmers.jpaboard.domain.user.entity.User;
+
+class PostTest {
```
저는 entity test를 해본적 없어서 어떻게 해야할 지 막막하던데, 이런식으로 하는건 정말 멋진것 같아요 배워갑니ㅏㄷ

park-se-jun commented on src/test/java/com/programmers/jpaboard/domain/post/repository/PostRepositoryTest.java:
```diff
@@ -0,0 +1,83 @@
+import com.programmers.jpaboard.domain.user.repository.UserRepository;
+
+@DataJpaTest
+class PostRepositoryTest {
```
저도 페어프로그래밍 하면서 생긴 의문인데요, JPA 에서 알아서 Repository를 만들어주고, 그걸 따로 튜닝하지 않았는데 Repository의 테스트가 필요할까요? 어떻게 생각하나요?

Seongjun-Kwon replied:
저도 test 코드를 만들면서 생각한 부분인데, 저는 기본적인 crud 의 경우 필요하지 않다고 결론을 내렸습니다.
만약, repository 에 스스로 쿼리를 만들어서 넣어준 로직이 있다면 필요할 것 같습니다.

park-se-jun commented on src/test/java/com/programmers/jpaboard/domain/post/service/PostServiceTest.java:
```diff
@@ -0,0 +1,119 @@
+@ExtendWith(MockitoExtension.class)
+class PostServiceTest {
+
+	@InjectMocks
```
@Mock 객체를 넣은 인스턴스를 생성하는 어노테이션 배워갑니당~ 퍼가요~~😊

park-se-jun commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+				.contentType(MediaType.APPLICATION_JSON)
+				.accept(MediaType.APPLICATION_JSON)
+				.content(requestBody))
+			.andExpect(jsonPath("id").value(createdPostDto.getId()))
```
Response인 json 의 데이터를 andExpect 와 jsonPath로 검증할 수 있군요! 저 이걸 어떻게 해야할지 몰랐는데 멋져요👍👍

park-se-jun commented on src/test/java/com/programmers/jpaboard/web/post/PostApiControllerTest.java:
```diff
@@ -0,0 +1,179 @@
+	}
+
+	@Test
+	void getPosts() throws Exception {
```
문서화 되지 않는 get /posts .... 프론트엔드 죽어욧~~

Seongjun-Kwon replied:
추가했습니다!

park-se-jun commented on src/docs/asciidoc/index.adoc:
```diff
@@ -0,0 +1,50 @@
+include::{snippets}/post-save/response-fields.adoc[]
+
+=== 게시판 조회
+
```
GET /api/v1/posts 도 해주세요 ㅠㅠㅠㅠ

Seongjun-Kwon replied:
추가했습니다!

park-se-jun reviewed (commented):
마르코팀의 거인(물리적) 성준이형 리뷰 다했슴당

Seongjun-Kwon committed:
refactor: DTO 의 기본생성자 관련 수정.

- 외부 생성 막기 위해 접근제한자 protected 로 변경.
- response DTO 에 불필요한 기본생성자 제거.

Seongjun-Kwon committed:
refactor: static 메서드만 가지고 있는 Converter 들의 생성자를 private 로 변경.

Seongjun-Kwon committed:
refactor: Post 생성자에 User 필수적으로 넣음.

- Post 는 User 에 종속적이기 때문.

Seongjun-Kwon committed:
refactor: User 엔티티의 posts 필드에 설정된 cascade 속성 변경.

- 기존 ALL -> PERSIST 로 변경.

Seongjun-Kwon committed:
refactor: 컨트롤러에서 다중 게시글 조회 시의 반환 값 변경.

- 기존에 Page<PostResponseDto> 를 반환했었음.
- List<PostResponseDto> 를 필드로 갖는 PostResponseDtoList 를 반환하도록 변경.

Seongjun-Kwon committed:
refactor: 405 에러 예외 처리 추가.

ksy90101 commented on src/main/java/com/programmers/jpaboard/common/exception/UserNotFoundException.java:
```diff
@@ -0,0 +1,10 @@
+
+import javax.persistence.EntityNotFoundException;
+
+public class UserNotFoundException extends EntityNotFoundException {
```
한번 생각해보면 좋을거 같아요.
현재는 Entity가 몇개 없지만, Entity가 많아짐에 따라 이렇게 예외를 만드는것이 과연 관리측면에서 좋을지?
EntityNotFoundException만으로도 충분히 의미 전달이 가능하지 않을까?? 라는 생각을 해보면 좋을거 같아요.

Seongjun-Kwon replied:
말씀해주신대로 현재는 엔티티가 몇 개 없지만, 프로젝트가 커질수록 이렇게 세분화한다면 관리해줘야할 예외들이 굉장히 많아질 것 같습니다. EntityNotFoundException 하나로 추상화하여도 충분히 의미 전달이 가능할 것 같습니다.
서버에서 어떤 엔티티에서 예외가 터졌는지를 알고 싶다면 로그를 이용하는게 좋아보입니다.

Seongjun-Kwon committed:
docs: REST DOCS 문서 추가.

- 게시글 페이지 조회 추가.
- 사용자 생성 추가.

Seongjun-Kwon committed:
refactor: 201(CREATED) 응답 시 생성된 자원의 uri 를 반환하게 함.

ksy90101 reviewed (approved):
전반적으로 너무 잘 해주셨네요 👍 
JPA 게시판은 마무리 하겠습니다.
리뷰 하나는 고민해보세요~
